### PR TITLE
Infoblox

### PR DESCRIPTION
**Pull request by jgrassler**
This pull request adds support for the [new style infoblox Neutron plugin](https://github.com/openstack/networking-infoblox/).

### CODE REVIEW

**Comment by jgrassler** [comment]
This is now ready for review. I tested it manually up to the point of provoking the following exception in `/var/log/neutron/neutron-server.log`

```
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource Traceback (most recent call last):
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource   File "/usr/lib/python2.7/site-packages/neutron/api/v2/resource.py", line 84, in resource
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource     result = method(request=request, **args)
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource   File "/usr/lib/python2.7/site-packages/neutron/api/v2/base.py", line 410, in create
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource     return self._create(request, body, **kwargs)
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource   File "/usr/lib/python2.7/site-packages/oslo_db/api.py", line 148, in wrapper
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource     ectxt.value = e.inner_exc
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource   File "/usr/lib/python2.7/site-packages/oslo_utils/excutils.py", line 220, in __exit__
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource     self.force_reraise()
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource   File "/usr/lib/python2.7/site-packages/oslo_utils/excutils.py", line 196, in force_reraise
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource     six.reraise(self.type_, self.value, self.tb)
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource   File "/usr/lib/python2.7/site-packages/oslo_db/api.py", line 138, in wrapper
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource     return f(*args, **kwargs)
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource   File "/usr/lib/python2.7/site-packages/neutron/api/v2/base.py", line 521, in _create
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource     obj = do_create(body)
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource   File "/usr/lib/python2.7/site-packages/neutron/api/v2/base.py", line 503, in do_create
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource     request.context, reservation.reservation_id)
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource   File "/usr/lib/python2.7/site-packages/oslo_utils/excutils.py", line 220, in __exit__
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource     self.force_reraise()
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource   File "/usr/lib/python2.7/site-packages/oslo_utils/excutils.py", line 196, in force_reraise
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource     six.reraise(self.type_, self.value, self.tb)
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource   File "/usr/lib/python2.7/site-packages/neutron/api/v2/base.py", line 496, in do_create
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource     return obj_creator(request.context, **kwargs)
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource   File "/usr/lib/python2.7/site-packages/neutron/plugins/ml2/plugin.py", line 892, in create_subnet
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource     result, mech_context = self._create_subnet_db(context, subnet)
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource   File "/usr/lib/python2.7/site-packages/neutron/plugins/ml2/plugin.py", line 881, in _create_subnet_db
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource     result = super(Ml2Plugin, self).create_subnet(context, subnet)
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource   File "/usr/lib/python2.7/site-packages/neutron/db/db_base_plugin_v2.py", line 703, in create_subnet
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource     return self._create_subnet(context, subnet, subnetpool_id)
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource   File "/usr/lib/python2.7/site-packages/neutron/db/db_base_plugin_v2.py", line 594, in _create_subnet
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource     subnetpool_id)
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource   File "/usr/lib/python2.7/site-packages/neutron/db/ipam_pluggable_backend.py", line 434, in allocate_subnet
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource     ipam_driver = driver.Pool.get_instance(subnetpool, context)
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource   File "/usr/lib/python2.7/site-packages/neutron/ipam/driver.py", line 54, in get_instance
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource     return driver_class(subnet_pool, context)
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource   File "/usr/lib/python2.7/site-packages/networking_infoblox/ipam/driver.py", line 51, in func
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource     return f(*args, **kwargs)
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource   File "/usr/lib/python2.7/site-packages/networking_infoblox/ipam/driver.py", line 91, in __init__
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource     self._grid_manager.get_config()
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource   File "/usr/lib/python2.7/site-packages/networking_infoblox/neutron/common/grid.py", line 85, in get_config
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource     self.grid_config.sync()
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource   File "/usr/lib/python2.7/site-packages/networking_infoblox/neutron/common/grid.py", line 244, in sync
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource     discovered_config = self._discover_config(self._get_gm_member())
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource   File "/usr/lib/python2.7/site-packages/networking_infoblox/neutron/common/grid.py", line 305, in _get_gm_member
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource     raise exc.InfobloxCannotFindMember(member="GM")
2016-09-12 11:43:57.868 18138 ERROR neutron.api.v2.resource InfobloxCannotFindMember: Cannot find the member: 'GM'
```

This means the infoblox-ipam-agent is running and configured in `neutron.conf` but cannot connect to the bogus Infoblox grid master I configured. This is about as far as we can test things without an actual Infoblox API to play with. Testing this is a bit tricky right now since packages are only in my home project and/or Cloud:OpenStack:Master (some dependencies) so far. Steps for testing:

1. Build a cloud using `mkcloud instonly`
2. Apply this commit's changes
3. Apply barclamps up to the Neutron barclamp
4. Add repositories:
  * zypper ar http://download.opensuse.org/repositories/home:/jgrassler:/branches:/Cloud:/OpenStack:/Master/SLE_12_SP1/ jgr-master
  * zypper ar http://download.opensuse.org/repositories/Cloud:/OpenStack:/Master/SLE_12_SP1/ com
5. Set `use_infoblox` to `true` in Neutron barclamp and deploy Neutron barclamp
6. Deploy Nova barclamp

Note: the Nova barclamp needed a little change as well because `infoblox-ipam-agent` needs Nova to use the `messagingv2` driver for RabbitMQ (see https://github.com/openstack/networking-infoblox/blob/master/doc/source/installation.rst#nova).

In the (unlikely) case that there are no further comments: please do not merge this right away, I'd like to rebase this mess of micro commits before this happens, but I'd like to retain the history just in case I need to revert/revisit stuff.

File: chef/cookbooks/neutron/attributes/default.rb [hunk 1]
```diff
@@ -83,6 +83,7 @@
     zvm_agent_name: "openstack-neutron-zvm-agent",
     lbaas_agent_pkg: "openstack-neutron-lbaas-agent",
     lbaas_agent_name: "openstack-neutron-lbaas-agent",
+    infoblox_agent_name: "openstack-neutron-infoblox-ipam-agent",
```
Comment (vuntz) [inline chef/cookbooks/neutron/attributes/default.rb]: Can you define this for the other distros too?

File: chef/cookbooks/neutron/libraries/helpers.rb [hunk 2]
```diff
@@ -67,4 +67,44 @@ def self.get_network_node_from_neutron_attributes(node)
       Chef::Node.load(network_node_name)
     end
   end
+
+  # Validates infoblox configuration and returns a configuration hash ready for
+  # passing to the neutron.conf template.
+  def self.get_infoblox_config(node)
+    return { use_infoblox: false } unless node[:neutron][:use_infoblox]
```
Comment (vuntz) [inline chef/cookbooks/neutron/libraries/helpers.rb]: I would just call the attribute `enabled` since we know the hash we'll get is about Infoblox.

File: chef/cookbooks/neutron/libraries/helpers.rb [hunk 3]
```diff
@@ -67,4 +67,44 @@ def self.get_network_node_from_neutron_attributes(node)
...
+    node[:neutron][:infoblox][:use_infoblox] = node[:neutron][:use_infoblox]
```
Comment (vuntz) [inline chef/cookbooks/neutron/libraries/helpers.rb]: Doesnt't reall matter, but I'd put `= true` directly here.

File: chef/cookbooks/neutron/libraries/helpers.rb [hunk 4]
```diff
@@ -67,4 +67,44 @@ def self.get_network_node_from_neutron_attributes(node)
...
+      raise "No grid configurations in node[:neutron][:infoblox][:grids]. " \
+            "Please specify at least one."
+    end
+
+    if node[:neutron][:infoblox][:cloud_data_center_id].to_i >=
+        node[:neutron][:infoblox][:grids].length
+      raise "Invalid cloud_data_center_id: " \
+            "`#{node[:neutron][:infoblox][:cloud_data_center_id]}` is not " \
+            "a valid index into node[:neutron][:infoblox][:grids]."
+    end
```
Comment (vuntz) [inline chef/cookbooks/neutron/libraries/helpers.rb]: Can we also have these checks in the validation of `neutron_service.rb`?

File: chef/cookbooks/neutron/recipes/common_config.rb [hunk 5]
```diff
@@ -72,6 +72,8 @@
 
 keystone_settings = KeystoneHelper.keystone_settings(neutron, @cookbook_name)
```
Comment (vuntz) [inline chef/cookbooks/neutron/recipes/common_config.rb]: Personal preference: I would not have the empty line here.

File: chef/cookbooks/neutron/recipes/common_config.rb [hunk 6]
```diff
@@ -116,6 +118,9 @@
 os_sdn_net = Barclamp::Inventory.get_network_definition(node, "os_sdn")
 mtu_value = os_sdn_net.nil? ? 1500 : os_sdn_net["mtu"].to_i
 
+ipam_driver = nil
+ipam_driver = "infoblox" if infoblox_settings[:use_infoblox]
```
Comment (vuntz) [inline chef/cookbooks/neutron/recipes/common_config.rb]: Why not `ipam_driver = infoblox_settings[:use_infoblox] ? "infoblox" : nil`?

File: chef/cookbooks/neutron/recipes/server_ha.rb [hunk 7]
```diff
@@ -31,23 +31,24 @@
...
-pacemaker_primitive primitive_name do
-  agent node[:neutron][:ha][:server][:server_ra]
-  op node[:neutron][:ha][:server][:op]
-  action :update
-  only_if { CrowbarPacemakerHelper.is_cluster_founder?(node) }
-end
-transaction_objects << "pacemaker_primitive[#{primitive_name}]"
+["neutron-server", "infoblox"].each do |primitive_name|
+  pacemaker_primitive primitive_name do
+    agent node[:neutron][:ha][primitive_name.to_sym]["#{primitive_name}_ra".to_sym]
```
Comment (vuntz) [inline chef/cookbooks/neutron/recipes/server_ha.rb]: You're breaking stuff for neutron-server here as the attribute is `:server`, not `:neutron-server`.

Comment (vuntz) [inline chef/cookbooks/neutron/recipes/server_ha.rb]: Also, I would probably name the primitive for the infoblox bit `neutron-infoblox-ipam-agent`, for consistency with the other names we have.

File: chef/cookbooks/neutron/recipes/server.rb [hunk 8]
```diff
@@ -370,6 +372,15 @@
   provider Chef::Provider::CrowbarPacemakerService if ha_enabled
 end
 
+if node[:neutron][:use_infoblox]
+  service node[:neutron][:platform][:infoblox_agent_name] do
```
Comment (vuntz) [inline chef/cookbooks/neutron/recipes/server.rb]: Just to confirm: this agent should be running on the same node as neutron-server, not on the same node as other network agents. Is that correct?

File: chef/cookbooks/neutron/recipes/server_ha.rb [hunk 9]
```diff
@@ -31,23 +31,24 @@
...
+  end
+  transaction_objects << "pacemaker_primitive[#{primitive_name}]"
 
-clone_name = "cl-#{primitive_name}"
-pacemaker_clone clone_name do
-  rsc primitive_name
-  meta ({ "clone-max" => CrowbarPacemakerHelper.num_corosync_nodes(node) })
-  action :update
-  only_if { CrowbarPacemakerHelper.is_cluster_founder?(node) }
+  clone_name = "cl-#{primitive_name}"
```
Comment (vuntz) [inline chef/cookbooks/neutron/recipes/server_ha.rb]: Are we sure the infoblox ipam agent can run active/active?

File: chef/cookbooks/neutron/templates/default/neutron.conf.erb [hunk 10]
```diff
@@ -163,6 +163,9 @@ allow_overlapping_ips = False
 # (default behavior), no IPAM driver is used. In order to use the reference
 # implementation of Neutron IPAM driver, use 'internal'. (string value)
 #ipam_driver = <None>
+<% unless @ipam_driver.nil? -%>
+ipam_driver=<%= @ipam_driver %>
```
Comment (vuntz) [inline chef/cookbooks/neutron/templates/default/neutron.conf.erb]: Spaces around `=`?

File: chef/cookbooks/neutron/libraries/helpers.rb [hunk 11]
```diff
@@ -67,4 +67,44 @@ def self.get_network_node_from_neutron_attributes(node)
...
+                   :ssl_verify,
+                   :wapi_max_results,
+                   :wapi_version]
+
+    # Fill in defaults for individual grid configurations ([infoblox-dc:<id>])
+
+    node[:neutron][:infoblox][:grids].each do |grid|
+      defaultable.each do |d|
+        unless grid.key?(d)
+          grid[d] = node[:neutron][:infoblox][:grid_defaults][d]
```
Comment (vuntz) [inline chef/cookbooks/neutron/libraries/helpers.rb]: Hrm. Not a big fan of this, for a couple of reasons:

- this sets an attribute on the node, so you should either use `node.set` or `node.default` or something like this
- generally speaking, we should avoid setting attributes in the namespace of a barclamp on a node: that is, we should avoid setting stuff in `node[:nova]`, `node[:keystone]`, etc. That's because the attributes in this namespace are actually note on the node itself, but inherited from a role the node has, and if we start adding them on the node, it will possibly be confusing, cause bugs (because of unclear priority between role attribute and node attribute), and so on.

I would probably just kill that code and instead, in the template, do something like: `option = grid_option.nil? grid_default : grid_option`

File: chef/cookbooks/neutron/templates/default/neutron.conf.erb [hunk 12]
```diff
@@ -1594,3 +1597,73 @@ key_file = <%= @ssl_key_file %>
...
+
+# HTTP pool max size.
+# http_pool_maxsize = 100
+http_pool_maxsize = <%= @infoblox[:grids][grid][:http_pool_maxsize] %>
+
+# HTTP request timeout.
+# http_request_timeout = 120
+http_request_timeout = <%= @infoblox[:grids][grid][:http_request_timeout] %>
+<% end -%>
+
```
Comment (vuntz) [inline chef/cookbooks/neutron/templates/default/neutron.conf.erb]: Empty line should be before the `<% end -%>` bit, to have an empty line between two sections.

File: crowbar_framework/app/models/nova_service.rb [hunk 13]
```diff
@@ -176,6 +176,13 @@ def create_proposal
     base["attributes"]["nova"]["db"]["password"] = random_password
     base["attributes"]["nova"]["neutron_metadata_proxy_shared_secret"] = random_password
 
+    neutron = Proposal.find_by(barclamp: "neutron",
+                               name: base["attributes"][@bc_name]["neutron_instance"])
+
+    # Use messagingv2 if the infoblox agent is enabled for the Neutron barclamp
+    use_infoblox = neutron["attributes"]["neutron"]["use_infoblox"]
+    base[:attributes][:nova][:notification_driver] = "messagingv2" if use_infoblox
```
Comment (vuntz) [inline crowbar_framework/app/models/nova_service.rb]: Can't we just always switch to messagingv2 instead of adding an attribute?

File: chef/data_bags/crowbar/migrate/neutron/102_infoblox_update.rb [hunk 14]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+def upgrade(ta, td, a, d)
+  unless a.key? "use_infoblox"
+    a["use_infoblox"] = ta["use_infoblox"]
+  end
+  # Since we will backport this to Cloud 6, we use, 'grid_defaults`, one of the
```
Comment (vuntz) [inline chef/data_bags/crowbar/migrate/neutron/102_infoblox_update.rb]: "we use," => "we use"?

File: chef/cookbooks/neutron/attributes/default.rb (hunk 1 above)
Comment (jgrassler) [inline chef/cookbooks/neutron/attributes/default.rb]: I can, but not sure if it's a good idea: right now there are no packages as far as I can see, so that name might change (e.g. the RHEL package might end up getting named `openstack-neutron-infoblox-agent` rather than `openstack-neutron-infoblox-ipam-agent`.

Comment (vuntz) [inline chef/cookbooks/neutron/attributes/default.rb]: Sure, just make the thing empty. We do that for some other agents already.

File: chef/cookbooks/neutron/recipes/common_config.rb (hunk 6 above)
Comment (jgrassler) [inline chef/cookbooks/neutron/recipes/common_config.rb]: I don't like the `? :` operator very much - it is easy to overlook and hence almost invariably slows at least me down when reading the code. I.e. this is mostly a matter of personal preference :-)

If you'd rather have a one liner here I can change it, though.

File: chef/cookbooks/neutron/recipes/server.rb (hunk 8 above)
Comment (jgrassler) [inline chef/cookbooks/neutron/recipes/server.rb]: Correct, yes. Its task is providing `neutron-server` with IP addresses and other network configuration it retrieves from the Infoblox appliance.

File: chef/cookbooks/neutron/recipes/server_ha.rb (hunk 9 above)
Comment (jgrassler) [inline chef/cookbooks/neutron/recipes/server_ha.rb]: Reasonably sure... the vendor claims that is the case: https://github.com/openstack/networking-infoblox/blob/master/doc/source/installation.rst#installing-the-driver

Comment (vuntz) [inline chef/cookbooks/neutron/recipes/server_ha.rb]: Thanks for the confirmation :-)

File: chef/cookbooks/neutron/recipes/common_config.rb (hunk 6 above)
Comment (vuntz) [inline chef/cookbooks/neutron/recipes/common_config.rb]: Nah, I'm fine. It's a matter of taste, so doesn't matter.

File: chef/cookbooks/neutron/libraries/helpers.rb (hunk 11 above)
Comment (jgrassler) [inline chef/cookbooks/neutron/libraries/helpers.rb]: Hmm. Sounds to me like we could improve matters greatly by moving all code modifying stuff to `crowbar_framework/app/models/neutron_server.rb`. This would make the whole thing more transparent too. Once that's done I can get rid of get_infoblox_config entirely (by that time there won't be much left anyway :-)).

Comment (vuntz) [inline chef/cookbooks/neutron/libraries/helpers.rb]: That makes sense!

File: crowbar_framework/app/models/nova_service.rb (hunk 13 above)
Comment (jgrassler) [inline crowbar_framework/app/models/nova_service.rb]: We could (we do that kind of thing in the tempest barclamp for instance to match Heat's `heat_stack_owner_roles setting`), but I went for an attribute for transparency's sake: that way the user can easily see that Nova switched to `messagingv2`.

File: chef/cookbooks/neutron/libraries/helpers.rb (hunk 11 above)
Comment (jgrassler) [inline chef/cookbooks/neutron/libraries/helpers.rb]: Ok, done :-)

While I was at it I shortened it a bit, too since it doesn't really make sense to have a list of defaults in the code when it's already in the data bag. Also, I'm not quite sure what I'm doing in `neutron_server.rb` makes sense: are the contents of the data bag already available in `base[:attributes]` when `create_proposal()` runs?

This is in a separate commit as well so I can revisit it for amending.

File: chef/data_bags/crowbar/migrate/neutron/102_infoblox_update.rb (hunk 14 above)
Comment (jgrassler) [inline chef/data_bags/crowbar/migrate/neutron/102_infoblox_update.rb]: Fixed.

File: chef/cookbooks/neutron/templates/default/neutron.conf.erb (hunk 12 above)
Comment (jgrassler) [inline chef/cookbooks/neutron/templates/default/neutron.conf.erb]: Done.

File: chef/cookbooks/neutron/templates/default/neutron.conf.erb (hunk 10 above)
Comment (jgrassler) [inline chef/cookbooks/neutron/templates/default/neutron.conf.erb]: Fixed.

File: chef/cookbooks/neutron/recipes/server_ha.rb (hunk 7 above)
Comment (jgrassler) [inline chef/cookbooks/neutron/recipes/server_ha.rb]: Yeah. I was toying with the idea of `not` making this a loop. I guess i should have stuck with that idea :-)

I gave it a bit of an overhaul now: I modeled it on the way Magnum does HA (i.e. I created a group/clone group). I put this into a separate commit in case it needs revisiting.

File: chef/cookbooks/neutron/recipes/common_config.rb (hunk 5 above)
Comment (jgrassler) [inline chef/cookbooks/neutron/recipes/common_config.rb]: No longer relevant anyway, `get_infoblox_settings()` being a thing of the past now.

File: chef/cookbooks/neutron/libraries/helpers.rb (hunk 4 above)
Comment (jgrassler) [inline chef/cookbooks/neutron/libraries/helpers.rb]: Sure. In fact, I moved them over completely since they don't serve much of a purpose in the helper once they're in `neutron_service.rb`.

File: chef/cookbooks/neutron/libraries/helpers.rb (hunk 3 above)
Comment (jgrassler) [inline chef/cookbooks/neutron/libraries/helpers.rb]: Agreed, that makes the code a bit easier on the eyes. Done.

File: chef/cookbooks/neutron/libraries/helpers.rb (hunk 2 above)
Comment (jgrassler) [inline chef/cookbooks/neutron/libraries/helpers.rb]: Ok, done.

File: chef/cookbooks/neutron/attributes/default.rb (hunk 1 above)
Comment (jgrassler) [inline chef/cookbooks/neutron/attributes/default.rb]: Ok, done.

**Comment by jgrassler** [comment]
Ok, the last 3 commits hopefully fix all bugs I created while addressing the comments. So this is ready for another round of reviews now.

File: crowbar_framework/app/models/nova_service.rb (hunk 13 above)
Comment (dirkmueller) [inline crowbar_framework/app/models/nova_service.rb]: I think that should be transparent to the user (read: he shouldn't care). iirc messagingv2 is the recommended choice meanwhile, we should just always use it.

File: crowbar_framework/config/locales/neutron/en.yml [hunk 15]
```diff
@@ -105,6 +105,8 @@ en:
         no_valid_ip: 'Multicast group for VXLAN broadcast emulation %{mcast_group} is not a valid IP address'
         no_valid_multicast_ip: 'Multicast group for VXLAN broadcast emulation %{mcast_group} is not a valid multicast IP address.'
         network: 'Network "%{ext_net}" cannot be used as an additional external network'
+        infoblox_grids: "No grid configurations in infoblox[:grids]. Please specify at least one."
+        infoblox_dc_id: "Invalid cloud_data_center_id: `%{dc_id}` is not a valid index into infoblox[:grids]."
```
Comment (dirkmueller) [inline crowbar_framework/config/locales/neutron/en.yml]: '%{dc_id}' needs to be a index in the range (1-x) into infoblox[:grids]

File: crowbar_framework/config/locales/neutron/en.yml [hunk 16]
```diff
@@ -105,6 +105,8 @@ en:
         no_valid_ip: 'Multicast group for VXLAN broadcast emulation %{mcast_group} is not a valid IP address'
         no_valid_multicast_ip: 'Multicast group for VXLAN broadcast emulation %{mcast_group} is not a valid multicast IP address.'
         network: 'Network "%{ext_net}" cannot be used as an additional external network'
+        infoblox_grids: "No grid configurations in infoblox[:grids]. Please specify at least one."
```
Comment (dirkmueller) [inline crowbar_framework/config/locales/neutron/en.yml]: Please specificy at least one grid configuration in infoblox[:grids]

File: crowbar_framework/app/models/nova_service.rb (hunk 13 above)
Comment (jgrassler) [inline crowbar_framework/app/models/nova_service.rb]: Ok, I made `messagingv2` the default now.

File: crowbar_framework/config/locales/neutron/en.yml (hunk 16 above)
Comment (jgrassler) [inline crowbar_framework/config/locales/neutron/en.yml]: Done.

File: crowbar_framework/config/locales/neutron/en.yml (hunk 15 above)
Comment (jgrassler) [inline crowbar_framework/config/locales/neutron/en.yml]: Done.

**Comment by jgrassler** [comment]
Heads-up: rebased this on current master (schema revision for Nova needed bumping).

File: chef/cookbooks/neutron/recipes/server_ha.rb [hunk 17]
```diff
@@ -29,21 +29,44 @@
...
+  only_if { CrowbarPacemakerHelper.is_cluster_founder?(node) }
+end
+transaction_objects << "pacemaker_group[#{group_name}]"
+
+clone_name = "cl-#{group_name}"
 pacemaker_clone clone_name do
-  rsc primitive_name
-  meta ({ "clone-max" => CrowbarPacemakerHelper.num_corosync_nodes(node) })
+  rsc group_name
+  meta("clone-max" => CrowbarPacemakerHelper.num_corosync_nodes(node))
```
Comment (vuntz) [inline chef/cookbooks/neutron/recipes/server_ha.rb]: Are you sure we can remove the `{}` bits here? If it's because of hound, I think it's a bug in hound.

File: crowbar_framework/app/models/neutron_service.rb [hunk 18]
```diff
@@ -99,6 +99,16 @@ def create_proposal
...
 
+    # Fill in missing defaults for infoblox grid configurations
+    base[:attributes][:neutron][:infoblox][:grids].each do |grid|
+      defaults = base["attributes"]["neutron"]["infoblox"]["grid_defaults"]
+      defaults.each_key.each do |d|
+        unless grid.key?(d)
+          grid[d] = defaults[d]
+        end
+      end
+    end
```
Comment (vuntz) [inline crowbar_framework/app/models/neutron_service.rb]: Hrm, I somehow doubt it should be in `create_proposal`, as this won't be done when the user modifies the proposal and adds a grid. Shouldn't it be in some `save` method instead?

File: crowbar_framework/app/models/neutron_service.rb [hunk 19]
```diff
@@ -286,6 +310,7 @@ def validate_proposal_after_save(proposal)
 
     validate_ml2(proposal) if plugin == "ml2"
     validate_dvr(proposal)
+    validate_infoblox(proposal)
```
Comment (vuntz) [inline crowbar_framework/app/models/neutron_service.rb]: Shouldn't we validate only if `use_infoblox` is `true`?

File: chef/data_bags/crowbar/template-nova.json [hunk 20]
```diff
@@ -10,6 +10,7 @@
       "glance_instance": "none",
       "cinder_instance": "none",
       "neutron_instance": "none",
+      "notification_driver": "messagingv2",
```
Comment (vuntz) [inline chef/data_bags/crowbar/template-nova.json]: I'm still not sold on the attribute here, as opposed to always setting this in the conf file. I just don't see the benefit of having it configurable -- is there one?

File: chef/cookbooks/neutron/recipes/server_ha.rb (hunk 17 above)
Comment (jgrassler) [inline chef/cookbooks/neutron/recipes/server_ha.rb]: It is because of hound. And it does actually work as it's supposed to (Ruby syntax allows you to omit the braces if the only parameter is a hash). That being said I don't really like it either. It is the sort of thing that causes every reader to stop and wonder (and write a snippet of sample code to test if this actually works in my case), all for the dubious value of slightly terser code. So I'm all for teaching hound not to enforce this :-)

File: chef/data_bags/crowbar/template-nova.json (hunk 20 above)
Comment (jgrassler) [inline chef/data_bags/crowbar/template-nova.json]: I originally used that so I could modify the Nova barclamp's messaging driver transparently based on the value of the Neutron barclamp's `use_infoblox` setting. So if you'd rather have this hardwired I won't argue about it. I only left it exposed in case a hypothetical somebody needs that for a hypothetical reason some time in the future, which is not much of a case for leaving it exposed :-)

Comment (vuntz) [inline chef/data_bags/crowbar/template-nova.json]: I'd go for hardcoding it, yes.

File: chef/data_bags/crowbar/template-neutron.json [hunk 21]
```diff
@@ -78,50 +78,34 @@
       },
       "use_infoblox": false,
```
Comment (vuntz) [inline chef/data_bags/crowbar/template-neutron.json]: Feel free to ignore, this just came to my mind during the monthly demo: we could clean up attributes and have `[:infoblox][:enabled]` instead of `[:use_infoblox]` and `[:infoblox]`.

File: crowbar_framework/app/models/neutron_service.rb (hunk 18 above)
Comment (jgrassler) [inline crowbar_framework/app/models/neutron_service.rb]: That sounds better, yes. I'm having some trouble finding a suitable `save` method to use, though, though: I tried overriding and then calling both `ServiceObject's` `save_proposal!()` or  `proposal_edit()` method both looked like likely candidates, but overriding and then calling them with the modified proposal didn't have any effect in either case. Is there another save method that's flying in under the radar somewhere?

Comment (jgrassler) [inline crowbar_framework/app/models/neutron_service.rb]: Never mind. I PEBKACed on that one by only applying the changes to my git checkout but not `/opt/dell` :-)

Works now.

File: crowbar_framework/app/models/neutron_service.rb (hunk 19 above)
Comment (jgrassler) [inline crowbar_framework/app/models/neutron_service.rb]: Indeed. Done.

File: chef/data_bags/crowbar/template-nova.json (hunk 20 above)
Comment (jgrassler) [inline chef/data_bags/crowbar/template-nova.json]: Ok, done.

File: chef/data_bags/crowbar/template-neutron.json (hunk 21 above)
Comment (jgrassler) [inline chef/data_bags/crowbar/template-neutron.json]: Yes, I'll ignore this for the time being. Thanks for the offer :-)

I'd rather do this cleanup in a follow-up pull request.

**Comment by jgrassler** [comment]
All comments addressed. I'm leaving the commits as they are for now. I'll rebase before merge.

File: chef/cookbooks/neutron/recipes/server_ha.rb [hunk 22]
```diff
@@ -29,21 +29,44 @@
...
+    op node[:neutron][:ha][:infoblox][:op]
+    action :update
+    only_if { CrowbarPacemakerHelper.is_cluster_founder?(node) }
+  end
+  transaction_objects << "pacemaker_primitive[#{infoblox_primitive_name}]"
+  primitives << infoblox_primitive_name
+end
+
+group_name = "g-neutron"
+
```
Comment (vuntz) [inline chef/cookbooks/neutron/recipes/server_ha.rb]: For cleanup: you can kill this empty line.

File: chef/cookbooks/neutron/recipes/server_ha.rb (hunk 17 above)
Comment (vuntz) [inline chef/cookbooks/neutron/recipes/server_ha.rb]: I'd really add the `{}` back, tbh, especially as we have it like this everywhere else. Can be in a cleanup PR.

**Comment by jgrassler** [comment]
Ok,  I removed the empty line and restored the braces (no way I'm going to wait until a cleanup PR if you're giving me carte blanche to spite Hound _now_). Let's see how that squash feature works out...

File: chef/cookbooks/neutron/recipes/server_ha.rb (hunk 22 above)
Comment (jgrassler) [inline chef/cookbooks/neutron/recipes/server_ha.rb]: Done.

**Comment by aplanas** [comment]
Can you use the pattern: `component: ...` for the commits that are not going to be squashed? Also the hound complain is easy to fix.

But LGTM basically (not tested)

**Comment by vuntz** [comment]
> Also the hound complain is easy to fix.

As discussed, we actually want to keep the code as it is.

**Comment by jgrassler** [comment]
| Can you use the pattern: component: ... for the commits that are not going to be squashed?

I am using it already: All commits after "neutron/nova: re-added infoblox support" will be squashed since I only left them in there temporarily (in case I needed to revisit them).

**Comment by rsalevsky** [comment]
Needs a rebase.

**Comment by jgrassler** [comment]
Rebase done: Squashed commit messages and rebased on current master (there was a small conflict in default.rb: some line numbers were off due to the addition of the lbaas stuff).

File: chef/cookbooks/neutron/recipes/server_ha.rb [hunk 23]
```diff
@@ -29,21 +29,43 @@
...
+  only_if { CrowbarPacemakerHelper.is_cluster_founder?(node) }
+end
+transaction_objects << "pacemaker_group[#{group_name}]"
+
+clone_name = "cl-#{group_name}"
 pacemaker_clone clone_name do
-  rsc primitive_name
-  meta ({ "clone-max" => CrowbarPacemakerHelper.num_corosync_nodes(node) })
+  rsc group_name
+  meta({"clone-max" => CrowbarPacemakerHelper.num_corosync_nodes(node)})
```
Comment (rsalevsky) [inline chef/cookbooks/neutron/recipes/server_ha.rb]: `meta{ "clone-max" =>CrowbarPacemakerHelper.num_corosync_nodes(node) }` should fix the hound issue

Comment (jgrassler) [inline chef/cookbooks/neutron/recipes/server_ha.rb]: It does (I had that exact fix in place at first), but we decided on overruling Hound on this because it makes the code look ambiguous needlessly. Unfortunately Github hid the discussion in the course of a force push, but here it is: https://github.com/crowbar/crowbar-openstack/pull/521#discussion_r81309533
